docs: new configuration guide (#233)
convox/convox: docs/tutorials/preparing-an-application.md

---
order: 1
---

# Preparing An Application

> In order to deploy an app on Convox you will need two things

Reviewer: ```suggestion
In order to deploy your app on Convox you will need two things
```

> * A Dockerfile

Reviewer: ```suggestion
* A `Dockerfile`
```

> * A convox.yml manifest

Reviewer: ```suggestion
* A `convox.yml` manifest
```

Resources are network-accessible external services. Most often, resources are databases. You can see the list of currently supported resources [here](..reference/primitives/app/resource#types). When you specify a resource, Convox will pull a public docker image for the resource type and run that resource in a container inside your cluster. Any service which is [linked](https://docs.convox.com/reference/primitives/app/resource#linking) to a resource will automatically have a connection string URL for that resource injected as an environment variable at runtime. Convox resources are durable and can be used for production but if you would prefer to use a cloud specific database service such as RDS or Cloud SQL you can do so using resource [overlays](https://docs.convox.com/reference/primitives/app/resource#overlays) which allow you to specify an external service to use in place of the containerized resource for specific environments. This can be a great cost savings technique if for example you want to run a containerized Postgres database in your dev and staging environments but use RDS in production without needing to make any code changes or special configuration.

### Services

> Services are the heart of your app. Services are scalable processes defined by a Dockerfile. Services can expose a port or set of ports and can be manual or automatically scaled. The most typical example of a service would be a web application such as a Ruby, Django, NodeJS or Go app that is accessed via an https connection. By default services are publically accessible but Convox also supports internal services that can only be accessed by other services running in the same Racks which can be useful for things like internal APIs. Convox also supports special use case services such as [agents](https://docs.convox.com/configuration/agents) and [singletons](https://docs.convox.com/reference/primitives/app/service). You can read the full specification for services [here](https://docs.convox.com/reference/primitives/app/service).

Reviewer: ```suggestion
Services are the heart of your app. Services are scalable processes defined by a Dockerfile. Services can expose a port or set of ports and can be manual or automatically scaled. The most typical example of a service would be a web application such as a Ruby, Django, NodeJS or Go app that is accessed via an https connection. By default services with ports are publicly accessible but Convox also supports internal services that can only be accessed by other services running in the same Racks which can be useful for things like internal APIs. Convox also supports special use case services such as [agents](https://docs.convox.com/configuration/agents) and [singletons](https://docs.convox.com/reference/primitives/app/service). You can read the full specification for services [here](https://docs.convox.com/reference/primitives/app/service).
```

```
environment:
  - DEVELOPMENT=True
resources:
  database:
    type: postgres
  broker:
    type: redis
services:
  web:
    build: .
    domain: ${DOMAIN}
    port: 8001
    resources:
      - database
      - broker
  worker:
    build: .
    command: celery -A web worker -l info
    resources:
      - database
      - broker
timers:
  reminderemails:
    schedule: "*/5 * * * *"
    command: python manage.py sendreminders
    service: worker
```

A few things to note here are that the web service specifies an internal port of 8001 which will be exposed externally on 443. The web service will also be made available on the domain(s) specified in the the DOMAIN environment variable and Convox will automatically provision a SSL [certificate](https://docs.convox.com/configuration/load-balancers#ssl-termination) for the specified domain. The worker service on the other hand does not expose any external ports. The worker service also uses a unique startup command while the web service uses the command specified within the Dockerfile. Hopefully this gives you a sense of what a typical convox.yml might look like for a production application.

> ## Going From Docker-Compose to convox.yml

Reviewer: ```suggestion
## Going From docker-compose.yml to convox.yml
```

> If you are already using `docker-compose.yml` for your application moving to a `conovx.yml` is a relatively straightforward process. As an example take a look at the following `docker-compose.yml`

Reviewer: ```suggestion
If you are already using `docker-compose.yml` for your application moving to a `convox.yml` is a relatively straightforward process. As an example take a look at the following `docker-compose.yml`
```

      - supportservice
      - redis
  web_secondary:
    image: myusername/privateimage:latest
    command: override_command.sh
    ports:
      - 3001
    links:
      - redis
  redis:
    image: "redis:alpine"
  supportservice:
    build:
      context: .
      dockerfile: Dockerfile.support
      links:
        - postgres
  postgres:
    image: postgres:12

```

So here we have 5 different services defined, one a httpd service, one a private image with an overridden command, one a plain Redis service, and one we're building locally from a custom Dockerfile, with a Postgres database image thrown in for good measure.

The `convox.yml` manifest for this app would look like

```
resources:
  redis:
    type: redis
  postgres:
    type: postgres
    options:
      version: 12
services:
  web:
    image: httpd
    port: 80
    volumes:
      - /tmp/something
    enviroment:
      - MY_ENVIRONMENT=development
    resources:
      - redis
  web_secondary:
    image: myusername/privateimage:latest
    command: override_command.sh
    port: 3001
    resources:
      - redis
  supportservice:
    build:
      path: .
      manifest: Dockerfile.support
    resources:
      - postgres
```

> Here you can see how we have turned the Redis and Postgres services into resources, which can be accessed by the other services. As we mentioned above by default those resources will be run in containers backed by durable storage but if you would prefer to use a cloud provider service you can do so using [resource overlays](https://docs.convox.com/reference/primitives/app/resource#overlays). We have also removed the links directive as services in a Rack are able to communicate each other using our built-in [service discovery](https://docs.convox.com/configuration/service-discovery). Otherwise, you will find the manifests to be nearly identical.

Reviewer: ```suggestion
Here you can see how we have turned the Redis and Postgres services into resources, which can be accessed by the other services. As we mentioned above, those resources will be run in containers backed by durable storage (by default) but if you would prefer to use a cloud provider service you can do so using [resource overlays](https://docs.convox.com/reference/primitives/app/resource#overlays). We have also removed the links directive as services in a Rack are able to communicate each other using our built-in [service discovery](https://docs.convox.com/configuration/service-discovery). Otherwise, you will find the manifests to be nearly identical.
```

## Next steps

> 

Reviewer: ```suggestion

Have a go at [Deploying an Application](deploying-an-application.md)
```
